libraries/AP_RPM/RPM_FFT.cpp

```diff
@@ -0,0 +1,119 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <AP_HAL/AP_HAL.h>
+#include <AP_AHRS/AP_AHRS.h>
+#include "RPM_FFT.h"
+
+extern const AP_HAL::HAL& hal;
+
+/* 
+   open the sensor in constructor
+*/
+AP_RPM_FFT::AP_RPM_FFT(AP_RPM &_ap_rpm, uint8_t _instance, AP_RPM::RPM_State &_state) :
+    AP_RPM_Backend(_ap_rpm, _instance, _state)
+{
+    instance = _instance;
+    hal.scheduler->register_timer_process(FUNCTOR_BIND_MEMBER(&AP_RPM_FFT::fast_timer_update, void));
+    hal.scheduler->register_io_process(FUNCTOR_BIND_MEMBER(&AP_RPM_FFT::slow_timer_update, void));
+    // limit max RPM to half buffer size
+    ap_rpm._maximum[instance].set_and_notify(MIN(60*RPM_FFT_WIDTH/2, ap_rpm._maximum[instance]));
+}
+
+
+/*
+  timer function called at 1kHz
+*/
+void AP_RPM_FFT::fast_timer_update(void)
+{
+    if (nsamples < ARRAY_SIZE(fft_buffer)) {
+        const AP_InertialSensor &ins = AP::ins();
+        uint32_t imu_sample_us = ins.get_last_update_usec();
+        if (imu_sample_us == last_imu_sample_us) {
+            // no new sample from INS
+            return;
+        }
+        last_imu_sample_us = imu_sample_us;
+
+        const Vector3f &accel = ins.get_accel(0);
```

Comment:
Should this be using the `&accel` we made earlier using `imu_instance`?

Reply:
I think the idea was to get the data at 400 hz not 1000hz.  But I may rethink that and be fine with the 1000hz rate.  see my comment below.